### PR TITLE
TiFlash late materialization

### PR DESCRIPTION
**Pull request by qiancai**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/13359
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tiflash/tiflash-late-materialization.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+# TiFlash Late Materialization
+
+> **Warning:**
+>
+> Currently, this is an experimental feature. The form and usage might be modified in future releases.
+
+This document describes how to use the TiFlash late materialization feature to accelerate queries in OLAP scenarios.
+
+By default, TiFlash reads all the data from the columns required by a query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to one of the filtering conditions, filters out the rows that meet the condition, and then scans the other column data of these rows for further calculation, thus reducing the amount of scan I/O and data parsing computation.
```
Comment (Lloyd-Pottiger) [inline tiflash/tiflash-late-materialization.md:14]:
```suggestion
By default, TiFlash reads all the data from the columns required by a query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the pushed down filtering conditions, finds the rows that meet the condition, and then scans the other column data of these rows for further calculation, thus reducing the amount of scan I/O and data processing computation.
```

`filters out` 是过滤掉，不准确。

File: tiflash/tiflash-late-materialization.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+If you want to improve the performance of certain queries in OLAP scenarios, you can enable the TiFlash late materialization feature at the session level or global level. By modifying the value of the [`tidb_opt_enable_late_materialization`](/system-variables.md#tidb_opt_enable_late_materialization-new-in-v700) system variable, you can choose to enable or disable the TiFlash delayed materialization feature.
+
+After enabling the TiFlash late materialization feature, the TiDB optimizer will determine which filtering condition will be pushed down based on statistics and the query's filtering conditions. The optimizer will prioritize pushing down the filtering conditions with high filtering rates. For detailed algorithms, see the [RFC document](https://github.com/pingcap/tidb/tree/master/docs/design/2022-12-06-support-late-materialization.md).
```
Comment (Lloyd-Pottiger) [inline tiflash/tiflash-late-materialization.md:18]:
```suggestion
After enabling the TiFlash late materialization feature, the TiDB optimizer will determine which filtering conditions will be pushed down based on statistics. The optimizer will prioritize pushing down the filtering conditions with low selectivity. For detailed algorithms, please check the [RFC document](https://github.com/pingcap/tidb/tree/master/docs/design/2022-12-06-support-late-materialization.md).
```

File: tiflash/tiflash-late-materialization.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
++-------------------------+----------+--------------+---------------+-------------------------------------------------------+
+| id                      | estRows  | task         | access object | operator info                                         |
++-------------------------+----------+--------------+---------------+-------------------------------------------------------+
+| TableReader_12          | 12288.00 | root         |               | MppVersion: 1, data:ExchangeSender_11                 |
+| └─ExchangeSender_11     | 12288.00 | mpp[tiflash] |               | ExchangeType: PassThrough                             |
+|   └─TableFullScan_9     | 12288.00 | mpp[tiflash] | table:t1      | pushed down filter:lt(test.t1.a, 1), keep order:false |
++-------------------------+----------+--------------+---------------+-------------------------------------------------------+
+```
+
+In this example, the `a < 1` filtering condition is pushed down to the TableScan operator. TiFlash first reads all data from column `a`, and then filters the rows that meet the `a < 1` condition. Next, TiFlash reads the data of columns `b` and `c` for these rows.
```
Comment (Lloyd-Pottiger) [inline tiflash/tiflash-late-materialization.md:36]:
```suggestion
In this example, the filtering condition `a < 1` is pushed down to the TableScan operator. TiFlash first reads all data from column `a`, and then finds the rows that meet the `a < 1` condition. Next, TiFlash reads the specific rows of columns `b` and `c`.
```

File: tiflash/tiflash-late-materialization.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+set session tidb_opt_enable_late_materialization=OFF;
+```
+
+```sql
+set global tidb_opt_enable_late_materialization=OFF;
+```
+
+## Implementation mechanism
+
+When a filtering condition is pushed down to the TableScan operator, the execution process of the TableScan operator mainly includes the following steps:
```
Comment (Lloyd-Pottiger) [inline tiflash/tiflash-late-materialization.md:94]:
```suggestion
When there are some filtering conditions are pushed down to the TableScan operator, the execution process of the TableScan operator mainly includes the following steps:
```

File: tiflash/tiflash-late-materialization.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+1. Reads the three columns `<handle, del_mark, version>`, performs multi-version concurrency control (MVCC) filtering, and then generates the MVCC Bitmap.
+2. Reads the columns related to the filtering condition, performs the filtering condition, and then generates the Filter Bitmap.
```
Comment (Lloyd-Pottiger) [inline tiflash/tiflash-late-materialization.md:97]:
```suggestion
2. Reads the columns related to the filtering conditions, performs the filtering conditions, and then generates the Filter Bitmap.
```

File: tiflash/tiflash-late-materialization.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+This document describes how to use the TiFlash late materialization feature to accelerate queries in OLAP scenarios.
+
+By default, TiFlash reads all the data from the columns required by a query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to one of the filtering conditions, filters out the rows that meet the condition, and then scans the other column data of these rows for further calculation, thus reducing the amount of scan I/O and data parsing computation.
+
+If you want to improve the performance of certain queries in OLAP scenarios, you can enable the TiFlash late materialization feature at the session level or global level. By modifying the value of the [`tidb_opt_enable_late_materialization`](/system-variables.md#tidb_opt_enable_late_materialization-new-in-v700) system variable, you can choose to enable or disable the TiFlash delayed materialization feature.
```
Comment (qiancai) [inline tiflash/tiflash-late-materialization.md:16]:
```suggestion
If you want to improve the performance of certain queries in OLAP scenarios, you can enable the TiFlash late materialization feature at the session level or global level. By modifying the value of the [`tidb_opt_enable_late_materialization`](/system-variables.md#tidb_opt_enable_late_materialization-new-in-v700) system variable, you can choose to enable or disable the TiFlash late materialization feature.
```

File: tiflash/tiflash-late-materialization.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+After enabling the TiFlash late materialization feature, the TiDB optimizer will determine which filtering condition will be pushed down based on statistics and the query's filtering conditions. The optimizer will prioritize pushing down the filtering conditions with high filtering rates. For detailed algorithms, see the [RFC document](https://github.com/pingcap/tidb/tree/master/docs/design/2022-12-06-support-late-materialization.md).
+
```
Comment (qiancai) [inline tiflash/tiflash-late-materialization.md:19]: "下推过滤率高"，英文对应 “low selectivity”？

File: tiflash/tiflash-late-materialization.md (hunk 2 above)
Comment (qiancai) [reply tiflash/tiflash-late-materialization.md:18]: "see" 是“参考”文档的通用用法，英文用户文档避免使用 please

File: tiflash/tiflash-late-materialization.md (hunk 7 above)
Comment (Lloyd-Pottiger) [reply tiflash/tiflash-late-materialization.md:19]: 高过滤率对应“low selectivity”

Comment (Lloyd-Pottiger) [reply tiflash/tiflash-late-materialization.md:19]: 或者 “high filtration rate”, 不是 “filtering rate”

**Review by Lloyd-Pottiger — APPROVED** [review]

File: system-variables.md [hunk 8]
```diff
@@ -3002,6 +3002,15 @@ mysql> desc select count(distinct a) from test.t;
...
 - This variable is used to control whether the optimizer estimates the number of rows based on column order correlation
 
+### tidb_opt_enable_late_materialization <span class="version-mark">New in v7.0.0</span>
+
+- Scope: SESSION | GLOBAL
+- Persists to cluster: Yes
+- Type: Boolean
+- Default value: `OFF`
+- Specifies whether to enable the [TiFlash late materialization](/tiflash/tiflash-late-materialization.md) feature.
+- By default, if a `SELECT` statement includes a filtering condition (`WHERE` clause), TiFlash scans all the data of the required columns before filtering. When this variable is set to `ON` to enable the TiFlash late materialization feature, TiFlash can first scan the column data related to the filtering conditions, filter the rows that meet the conditions, and then scan the data of other columns of these rows for further calculations, thereby reducing IO scans and computations of data processing.
```
Comment (hfxsd) [inline system-variables.md:3012]:
```suggestion
- By default, if a `SELECT` statement includes a filtering condition (`WHERE` clause), TiFlash scans all the data of the required columns before filtering. When this variable is set to `ON` to enable the TiFlash late materialization feature, TiFlash can first scan the column data related to the filtering conditions, filter the rows that meet the conditions, and then scan the data of other columns of these rows for further calculations, thereby reducing computations of IO scans and data processing.
```

File: tiflash/tiflash-late-materialization.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+# TiFlash Late Materialization
+
+> **Warning:**
+>
+> Currently, this is an experimental feature. The form and usage might be modified in future releases.
+
+This document describes how to use the TiFlash late materialization feature to accelerate queries in OLAP scenarios.
+
+By default, when receiving a query request, TiFlash reads all the data from the columns required by the query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the filtering conditions that are pushed down, filters the rows that meet the condition, and then scans the other column data of these rows for further calculation, thereby reducing IO scans and computations of data processing.
```
Comment (hfxsd) [inline tiflash/tiflash-late-materialization.md:14]:
```suggestion
By default, when receiving a query request, TiFlash reads all the data from the columns required by the query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the filtering conditions that are pushed down, filters the rows that meet the condition, and then scans the other column data of these rows for further calculation, thereby reducing computations of IO scans and data processing.
```

File: tiflash/tiflash-late-materialization.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+## Implementation mechanism
+
+When filtering conditions are pushed down to the TableScan operator, the execution process of the TableScan operator mainly includes the following steps:
+
+1. Reads the three columns `<handle, del_mark, version>`, performs multi-version concurrency control (MVCC) filtering, and then generates the MVCC Bitmap.
+2. Reads the columns related to the filtering conditions, filters the rows that meet the conditions, and then generates the Filter Bitmap.
+3. Performs an `AND` operation between the MVCC Bitmap and Filter Bitmap to generate the Final Bitmap.
+4. Reads the corresponding rows of the remaining columns according to the Final Bitmap.
+5. Returns the results.
```
Comment (hfxsd) [inline tiflash/tiflash-late-materialization.md:100]: 漏译了一句话。中文原文：
5. 合并两次读取的数据，返回结果。

**Review by hfxsd — APPROVED** [review]
Rest LGTM

File: system-variables.md (hunk 8 above)
Comment (Lloyd-Pottiger) [reply system-variables.md:3012]: IO 和 计算

File: tiflash/tiflash-late-materialization.md (hunk 10 above)
Comment (qiancai) [inline tiflash/tiflash-late-materialization.md:100]:
```suggestion
5. Merges the data read in steps 2 and 4, and then returns the results.
```

Comment (qiancai) [reply tiflash/tiflash-late-materialization.md]: added in https://github.com/pingcap/docs/pull/12939/commits/3e749f9cdf4788fea291c514fc5f1bade8ac9092

File: system-variables.md [hunk 11]
```diff
@@ -3002,6 +3002,15 @@ mysql> desc select count(distinct a) from test.t;
 - Default value: `ON`
 - This variable is used to control whether the optimizer estimates the number of rows based on column order correlation
 
+### tidb_opt_enable_late_materialization <span class="version-mark">New in v7.0.0</span>
+
+- Scope: SESSION | GLOBAL
+- Persists to cluster: Yes
+- Type: Boolean
+- Default value: `OFF`
+- Specifies whether to enable the [TiFlash late materialization](/tiflash/tiflash-late-materialization.md) feature.
```
Comment (qiancai) [inline system-variables.md:3011]:
```suggestion
- This variable is used to control whether to enable the [TiFlash late materialization](/tiflash/tiflash-late-materialization.md) feature.
```

**Pull request merged**